eclass/zig.eclass

```diff
@@ -0,0 +1,581 @@
+#
+# 	# If there are lazy dependencies that can be triggered together:
+# 	zig_live_fetch -Denable-wayland -Denable-xwayland
+#
+# 	# If there are 2 lazy dependencies that can't be triggered
+# 	# together in one call because they conflict:
+# 	zig_live_fetch -Dmain-backend=opengl
+# 	zig_live_fetch -Dmain-backend=vulkan
+# }
+# @CODE
+zig_live_fetch() {
+	# This function will likely be called in src_unpack,
+	# before [zig_]src_prepare, so this directory might not
+	# exist yet.
+	mkdir -p "${BUILD_DIR}" > /dev/null || die
+	pushd "${BUILD_DIR}" > /dev/null || die
+
+	local args=(
+		"${ZBS_ARGS_BASE[@]}"
+
+		--global-cache-dir "${ZBS_ECLASS_DIR}/"
+
+		# Function arguments
+		"${@}"
+	)
+
+	einfo "ZBS: live-fetching with:"
+	einfo "${args[@]}"
+	ezig build --fetch "${args[@]}"
+
+	popd > /dev/null || die
+}
+
+# @FUNCTION: zig_src_unpack
+# @DESCRIPTION:
+# Unpacks every archive in SRC_URI and ZBS_DEPENDENCIES,
+# in that order.
+zig_src_unpack() {
+	# Thanks to Alfred Wingate "parona" for inspiration here:
+	# https://gitlab.com/Parona/parona-overlay/-/blob/874dcfe03116574a33ed51f469cc993e98db1fa2/eclass/zig.eclass
+
+	if [[ "${#ZBS_DEPENDENCIES_SRC_URI}" -eq 0 ]]; then
+		default_src_unpack
+		return
+	fi
+
+	local zig_deps=()
+	for dependency in "${!ZBS_DEPENDENCIES[@]}"; do
+		zig_deps+=("${dependency}")
+	done
+
+	# First unpack non-Zig dependencies, so that
+	# tarball with all Git dependencies tarballs is unpacked early.
+	local dist
+	for dist in ${A}; do
+		if has "${dist}" "${zig_deps[@]}"; then
+			continue
+		fi
+		unpack "${dist}"
+	done
+
+	# Now unpack all Zig dependencies, including those that are
+	# now unpacked from tarball-tarball.
+	local zig_dep
+	for zig_dep in "${zig_deps[@]}"; do
+		# Hide now-spammy hash from stdout
+		ezig fetch --global-cache-dir "${ZBS_ECLASS_DIR}/" \
+			"${DISTDIR}/${zig_dep}" > /dev/null
+	done
+	einfo "ZBS: ${#zig_deps[@]} dependencies fetched"
+}
+
+# @FUNCTION: zig_src_prepare
+# @DESCRIPTION:
+# Calls default "src_prepare" function, creates BUILD_DIR directory
+# and enables or disables system mode (by adding to ZBS_BASE_ARGS),
+# depending on how many packages there are.
+#
+# System mode is toggled here and not in "src_unpack" because they
+# could have been fetched by "live_fetch" in live ebuilds instead.
+zig_src_prepare() {
+	default_src_prepare
+
+	mkdir -p "${BUILD_DIR}" || die
+	einfo "BUILD_DIR: \"${BUILD_DIR}\""
+
+	local system_dir="${ZBS_ECLASS_DIR}/p/"
+
+	# We are using directories.len instead of ZBS_DEPENDENCIES.len
+	# because ebuild might have fetched packages using live_fetch
+	# instead.
+	local -a packages=()
+	readarray -d '' -t packages < <(find "${system_dir}" -mindepth 1 \
+		-maxdepth 1 -type d -print0 2> /dev/null)
+	local count="${#packages[@]}"
+
+	if [[ "${count}" -gt 0 ]]; then
+		einfo "ZBS: system mode enabled, ${count} packages found"
+		ZBS_ARGS_BASE+=( --system "${system_dir}" )
+	else
+		einfo "ZBS: no packages found, no need to enable system mode"
```

Comment:
What's "system mode"? Does it do any harm to enable it unconditionally? Does, say, either option make us more likely to detect ebuild problems?

Reply:
But doesn't that imply you're actually enabling fetching unless `network-sandbox` is used? That's not fine at all.

Reply:
> unless `network-sandbox` is used? That's not fine at all.

Unconditionally enabling now... I guess author should pass `-fno-sys=[name1] -fno-sys=[name2]` manually if need to check completeness of ZBS_DEPENDENCIES.

35c458c71daa